### ISSUE TITLE
What to do when other API scripts require tokens to be selected?

### ISSUE THREAD

**Issue by phylll** [michael-buschbeck/mychs-macro-magic#204]
In implementing Midgard critical combat effects, there are a bunch of those where weapons are being lost. Those lost weapons tend to land on various grid cells around the token losing them. It's kinda neat to use `!Spawn` to make such weapons appear in the game:

```
!mmm   chat: !Spawn --name|tokenMuleWeapons --side|rand --offset|${offsetX},${offsetY} --order|top --controlledby|all --tokenProps|name:${ownerID.name}'s Weapon,showname:1
```

The problem is: the `SpawnDefaultToken` API script requires the reference token to be selected, which my MMM script obviously cannot guarantee. Trying to ensure this makes for pretty messy UX: extra text for the player/GM to read, to understand and to follow, plus the somewhat pointless click on that token and all the mistakes when people accidentally mix up their own and their opponent's tokens.

So: obviously this is just fun & games, even more so than everything we're doing. We could simply **not** spawn tokens for lost weapons at all.

If we want to do this, though, it would be great to have some solution. I can think of two:

1. Could MMM manipulate `selected` so that the `!Spawn` command would be fooled?
2. Is there a simple way to replace the `!Spawn` command for our uses (which seem much more limited than the many features the API script provides)?

**Comment by michael-buschbeck** [comment]
Fascinating!

**MMM modifying `selected`:** Conceivable, using the same crude mechanism already used to fix `playerid` for such API calls (#78). So if you'd have modified/overwritten `selected` in script code and do `!mmm chat: !foo`, MMM wouldn't just fudge `msg.playerid` but also `msg.selected` for the API call.

**MMM spawning tokens:** Shouldn't be too hard either. Now that we have structs (or lists of key-value pairs), passing initial properties for the token also shouldn't be an issue.

**Animating tokens with MMM:** Update token properties in a loop and use `delay()` for timing (and parallelism). Could even introduce an `!mmm background` ... `end background` block that runs in parallel to anything else that follows.

**Comment by michael-buschbeck** [comment]
> **MMM modifying `selected`:** Conceivable, using the same crude mechanism already used to fix `playerid` for such API calls (#78). So if you'd have modified/overwritten `selected` in script code and do `!mmm chat: !foo`, MMM wouldn't just fudge `msg.playerid` but also `msg.selected` for the API call.

Writing this rung a faint bell, so I checked: MMM already does that… _**almost**_.

That's to say it already fudges `selected` for API calls but it simply forwards the `selected` effective when the `!mmm chat:` command was queued.

One internal conceptual problem is how to pass the `selected` override stored in the runtime variable stash to the `chat()` function to pass on to the API call.

**Comment by phylll** [comment]
> **MMM spawning tokens:** Shouldn't be too hard either. Now that we have structs (or lists of key-value pairs), passing initial properties for the token also shouldn't be an issue.

This would have the advantage that you could just as easily provide something that would operate on absolute coordinates, so scripts would not require a reference token in regard to which they want to spawn another token. While the latter -- `SpawnDefaultToken`'s key limitation -- makes complete sense in my present use case (a token present on the current map loses its weapon) as well as a few other use cases that come to mind (a mage token conjures something related to their body, e.g. a circle of fire around them), there may well be other use cases in which there would usually not be a reference token present, such as the GM spawning a fireball for an enemy outside the current map. In such a case, sensible UX would simply require a target point to be provided, and not some complicated offsets from a mage's token position. Even when there is a reference token, there would simply be less math required if we had a way of spawning tokens *here* rather than *x cells to the right and y cells to the bottom of some other token*.

**Comment by michael-buschbeck** [comment]
> **MMM modifying `selected`:** Conceivable, using the same crude mechanism already used to fix `playerid` for such API calls (#78). So if you'd have modified/overwritten `selected` in script code and do `!mmm chat: !foo`, MMM wouldn't just fudge `msg.playerid` but also `msg.selected` for the API call.

So...

While the "MMM to support token spawning" discussion has plenty of merit, so does the notion of MMM making it possible to overwrite `selected` for nested API calls.

I've implemented that in an MMM 1.36.0 pre-release, installed in our test game.

You can set `selected` to any combination of token names, character names (with associated tokens), token IDs, or character IDs (with associated tokens), and that list will be passed on to nested API calls. If you set `selected` to `undef`, an empty list of selected tokens will be passed on, even if there are actually selected tokens.

You can also use this to _filter_ the user's selected tokens by saying `!mmm set selected = selected where ...HP < 5` (but you can't change the actual selection, of course).

**Comment by phylll** [comment]
Works very nicely for me in our test game :-)
